### PR TITLE
AL-5378 - layer styles supporting transparency

### PR DESCRIPTION
**Pull request by OlegRayskiy**

### CODE REVIEW

File: src/map/styles/AlloyLayerStyle.ts [hunk 1]
```diff
@@ -31,22 +31,30 @@ export class AlloyLayerStyle {
...
    * @param icon the icon font class name
    * @param labelMode the mode for displaying labels
+   * @param opacity the opacity of the style
    */
   constructor(
     styleId: string,
     colour: string,
     icon?: string,
     labelMode?: AlloyLayerStyleLabelMode,
+    opacity = 1,
```
Comment (cmcnicholas) [inline src/map/styles/AlloyLayerStyle.ts]: we shouldn't be using defaults, makes it notoriously difficult to refactor, pattern is to use `?` and assume default on the property instead e.g. `this.opacity = opacity ?? 1` also we should test opacity is between 0-1

File: src/map/styles/builders/AlloyCustomStyleBuilder.ts [hunk 2]
```diff
@@ -169,10 +169,16 @@ export class AlloyCustomStyleBuilder extends AlloyStyleBuilder<AlloyCustomFeatur
         radius,
         feature.properties.icon,
         colour,
+        1,
```
Comment (cmcnicholas) [inline src/map/styles/builders/AlloyCustomStyleBuilder.ts]: Might be worth switching this out with a constant, `LayerStyleOpacity.Opaque`

File: src/map/styles/utils/AlloyBallUtils.ts [hunk 3]
```diff
@@ -17,22 +17,24 @@ export abstract class AlloyBallUtils {
...
    * @param radius the radius of the ball
    * @param colour the colour of the ball
+   * @param opacity the opacity of the ball
    * @param geometryFunction the optional geometry function for the style
    */
   public static createBallStyle(
     radius: number,
-    colour: string,
+    colour: Colour,
+    opacity = 1,
```
Comment (cmcnicholas) [inline src/map/styles/utils/AlloyBallUtils.ts]: no defaults, these should be explicit

File: src/map/styles/utils/AlloyBallUtils.ts [hunk 4]
```diff
@@ -45,18 +47,20 @@ export abstract class AlloyBallUtils {
...
    *               multiplier)
    * @param colour the colour of the ball
+   * @param opacity the opacity of the ball
    * @param geometryFunction the optional geometry function for the style
    */
   public static createBallHaloStyle(
     radius: number,
-    colour: string,
+    colour: Colour,
+    opacity = 1,
```
Comment (cmcnicholas) [inline src/map/styles/utils/AlloyBallUtils.ts]: same as above

File: src/utils/ColourUtils.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ const DARKEN_AMOUNT = 0.08;
  */
 const LIGHTEN_AMOUNT = 0.16;
 
+export type Colour = string | [number, number, number, number];
```
Comment (cmcnicholas) [inline src/utils/ColourUtils.ts]: no jsdoc for this

File: src/map/styles/AlloyLayerStyleOpacity.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Wrapper for style opacity that limits opacity to 0-1 range
+ * @ignore
+ * @internal
```
Comment (cmcnicholas) [inline src/map/styles/AlloyLayerStyleOpacity.ts]:
```suggestion
```

I think these can be removed as these are meant to be public api?

File: src/map/styles/builders/AlloyClusterStyleBuilder.ts [hunk 7]
```diff
@@ -66,6 +67,7 @@ export class AlloyClusterStyleBuilder extends AlloyStyleBuilderWithLayerStyles<
       resolution,
       // icon is not in here because clusters don't have them
       layerStyle.colour,
+      state === AlloyStyleBuilderBuildState.Default ? layerStyle.opacity : 1,
```
Comment (cmcnicholas) [inline src/map/styles/builders/AlloyClusterStyleBuilder.ts]: does opacity here need to override `toString` ?

Comment (OlegRayskiy) [reply src/map/styles/builders/AlloyClusterStyleBuilder.ts]: Missed this one

File: src/map/styles/AlloyLayerStyleOpacity.ts (hunk 6 above)
Comment (OlegRayskiy) [reply src/map/styles/AlloyLayerStyleOpacity.ts]: I don't think they should be, layer style constructors still take numbers and wraps it with this to clamp the values

**Pull request merged**